<!DOCTYPE html> <html> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Spatial Omics II: Defining Multicellular Regions | Kamal Maher </title> <meta name="author" content="Kamal Maher"> <meta name="description" content="A principled approach to representing multicellular regions in spatial omics data"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%AB%A7&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://kmaherx.github.io/blog/2025/regions/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="https://cdn.jsdelivr.net/npm/img-comparison-slider@8.0.6/dist/styles.min.css" integrity="sha256-3qTIuuUWIFnnU3LpQMjqiXc0p09rvd0dmj+WkpQXSR8=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://cdn.jsdelivr.net/npm/swiper@11.0.5/swiper-bundle.min.css" integrity="sha256-yUoNxsvX+Vo8Trj3lZ/Y5ZBf8HlBFsB6Xwm7rH75/9E=" crossorigin="anonymous"> <script src="/assets/js/distillpub/template.v2.js"></script> <script src="/assets/js/distillpub/transforms.v2.js"></script> </head> <body> <d-front-matter> <script async type="text/json">
      {
            "title": "Spatial Omics II: Defining Multicellular Regions",
            "description": "A principled approach to representing multicellular regions in spatial omics data",
            "published": "May 18, 2025",
            "authors": [
              
              {
                "author": "Kamal Maher",
                "authorURL": "",
                "affiliations": [
                  {
                    "name": "MIT, Broad Institute",
                    "url": ""
                  }
                ]
              }
              
            ],
            "katex": {
              "delimiters": [
                {
                  "left": "$",
                  "right": "$",
                  "display": false
                },
                {
                  "left": "$$",
                  "right": "$$",
                  "display": true
                }
              ]
            }
          }
    </script> </d-front-matter> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Kamal</span> Maher </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="post distill"> <d-title> <h1>Spatial Omics II: Defining Multicellular Regions</h1> <p>A principled approach to representing multicellular regions in spatial omics data</p> </d-title> <d-byline></d-byline> <d-article> <d-contents> <nav class="l-text figcaption"> <h3>Contents</h3> <div> <a href="#introduction">Introduction</a> </div> <div> <a href="#simulated-data">Simulated data</a> </div> <ul> <li> <a href="#region-components">Region components</a> </li> <li> <a href="#region-gradients">Region gradients</a> </li> <li> <a href="#region-clusters">Region clusters</a> </li> </ul> <div> <a href="#mouse-brain-data">Mouse brain data</a> </div> <ul> <li> <a href="#region-components">Region components</a> </li> <li> <a href="#region-gradients">Region gradients</a> </li> <li> <a href="#region-clusters">Region clusters</a> </li> </ul> <div> <a href="#conclusion">Conclusion</a> </div> </nav> </d-contents> <style>.slider-with-shadows{--default-handle-shadow:0 0 5px rgba(0,0,0,1);--divider-shadow:0 0 5px rgba(0,0,0,0.5)}</style> <h2 id="introduction">Introduction</h2> <p>As we saw in the <a href="/blog/2025/graph-fourier/">last post</a>, low-pass filtering isolates large-scale patterns. Recall that low-pass filtering corresponds to multiplying our data matrix by a function of the graph Laplacian matrix. Let’s resume this thought by defining our variables so we are all roughly on the same page.</p> <p>Our gene expression is given by the cell-by-gene matrix</p> \[\mathbf{X} \in \mathbb{R}^{c \times g},\] <p>where $c$ is the number of cells and $g$ is the number of genes.</p> <p>One might think that performing conventional single-cell analysis on low-pass filtered data would identify large-scale transcriptional features over the tissue (as opposed to cell type features). In this section, we will show that this is indeed the case.</p> <p>First, we can visualize the result of performing low-pass filtering on a given gene expression signal over the tissue. Consider a marker gene for one of the inner regions of our tissue. Drag the slider left to right to compare the signal before and after filtering.</p> <div style="width: 50%; max-width: 768px; margin: 0 auto;"> <img-comparison-slider class="slider-with-shadows"> <figure slot="first"> <picture> <source class="responsive-img-srcset" srcset="/assets/figures/fourier/tissue_before_filtering-480.webp 480w,/assets/figures/fourier/tissue_before_filtering-800.webp 800w,/assets/figures/fourier/tissue_before_filtering-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/figures/fourier/tissue_before_filtering.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="lazy" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> <figure slot="second"> <picture> <source class="responsive-img-srcset" srcset="/assets/figures/fourier/tissue_after_lowpass-480.webp 480w,/assets/figures/fourier/tissue_after_lowpass-800.webp 800w,/assets/figures/fourier/tissue_after_lowpass-1400.webp 1400w," type="image/webp" sizes="95vw"></source> <img src="/assets/figures/fourier/tissue_after_lowpass.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="lazy" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </img-comparison-slider> </div> <figcaption><strong>Figure 1:</strong> Comparison of a gene expression signal before (left image) and after (right image) low-pass filtering. </figcaption> <p><br></p> <p>Note that the small-scale variation in expression has been washed away. Neighboring cells are forced to look more similar to one another, in effect isolating only <em>large</em>-scale patterns.</p> <hr> <h2 id="simulated-data">Simulated data</h2> <h3 id="region-components">Region components</h3> <p>Now that we’ve isolated large-scale gene expression patterns, we can start to <strong>compare them to find interesting gene-gene relationships</strong>. Intuitively, a group of gene expression patterns that overlap in the tissue should represent a region. We can find these groups by looking at pairwise relationships between gene signals. Consider low-pass filtered gene signals $\mathbf{\bar x}_i, \mathbf{\bar x}_j \in \mathbb{R}^n$. Because these signals are vectors and we want a scalar measure of similarity, we could just compare them by taking the inner product:</p> \[\mathbf{\bar x}_i^{\top} \mathbf{\bar x}_j \in \mathbb{R}.\] <p>With some mean centering, this is defined as the covariance. While we will continue to refer to this as covariance, we will omit all mean centering for simplicity (although it will often add a “junk component” as we will see in a moment).</p> <p>But we aren’t just interested in one pair of genes; <strong>we want to look at all pairwise relationships</strong>. Let $\mathbf{\bar X} = [\mathbf{\bar x}_1 | … | \mathbf{\bar x}_g] \in \mathbb{R}^{n \times g}$ represent the cell-by-gene matrix of low-pass filtered gene signals. Then the gene-by-gene covariance matrix is given by</p> \[\mathbf{C} = \mathbf{\bar X}^{\top} \mathbf{\bar X} \in \mathbb{R}^{g \times g},\] <p>with entries $\mathbf{C}_{ij} = \mathbf{\bar x}_i^{\top} \mathbf{\bar x}_j$. We can visualize $\mathbf{C}$ as a heatmap. The genes are sorted based on their corresponding ground truth region patterns, so we expect to see interesting groups of covarying genes as blocks. (If you look closely, you’ll see that these blocks are 4x4, as there are four gene markers for each pattern.) For instance, gene markers for each region form red blocks along the diagonal, denoting positively covarying groups of genes, i.e. gene programs. Furthermore, different blocks appear to negatively covary, forming blue blocks along the <em>off</em>-diagonal. Conceptually, this just means that gene markers for each region are mutually exclusive.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/region_covariance.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 2:</strong> Visualization of the low-pass covariance matrix.</figcaption> </figure> <p>Ultimately, <strong>we seek to distill these blocks into a simpler representation</strong> – perhaps by grouping genes within related blocks into a single “gene program” describing a region. It turns out we can do this by eigendecomposing $\mathbf{C}$, i.e. performing PCA. (For a primer on PCA, see <a href="/blog/2025/pca">this post</a>.) Briefly, eigendecomposing the gene-gene covariance matrix yields the eigenbasis</p> \[\mathbf{U} = [ \mathbf{u}_1 | ... | \mathbf{u}_g ] \in \mathbb{R}^{g \times g},\] <p>where $\mathbf{u}_i \in \mathbb{R}^g$ represents the gene loadings for PC$i$. In other words, it describes each gene’s participation in gene program $i$. We can visualize this matrix as a heatmap as well. The rows represent each gene in the same order as in Figure 2. The columns now represent PCs, or “region components”, and the colors describe how much and in what direction a given gene contributes to a given component.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/region_components.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 3:</strong> Visualization of low-pass gene programs, i.e. "region components". The right-hand plot is an enlarged version of the inset in the top left corner of the left-hand plot.</figcaption> </figure> <p>The first component is entirely negative and is likely just a consequence of neglecting mean centering (the “junk component” mentioned above). The second, third, and fourth components each appear to describe relationships between regions, with each representing a positive and negative region.</p> <p>We can also visualize each component in the tissue by projecting cells onto each gene program, i.e. $\mathbf{X} \mathbf{U} \in \mathbb{R}^{n \times g}$. We end up seeing that they each describe two regions of the tissue. The first component describes the outer ring versus everything inside of it, the second component describes the second-most outer right versus everything inside, and the third component describes the third-most outer ring versus everything inside. We can also sort gene loadings to identify gene markers for each region component. This gives a set of top genes describing the “positive region” and a different set for the “negative region”. We can visualize each of these markers in the tissue as well, seeing that they are indeed expressed within the regions they describe.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/region_components_tissue.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 4:</strong> Visualization of region components and their gene markers in the tissue.</figcaption> </figure> <p>Altogether, we find that <strong>each component represents a large-scale pattern between regions</strong>. Note that the gene marker information shown in Figure 4 is the same as the information shown in Figure 3, just in a different form.</p> <h3 id="region-gradients">Region gradients</h3> <p>The above components indeed appear to contain orthogonal information, as they each describe unique sets of regions. <strong>Together, however, these features might describe more complex region patterns</strong>. Rather than projecting cells onto a single component, one can instead project onto multiple components that form a multidimensional “region space”. In other words, rather than plotting cells in the tissue and <em>coloring</em> by the top three different region components as above, we can directly plot each cell in three-dimensional component space and see what shapes they form. Try interacting with the plot below to get a sense of the resulting shape.</p> <figure style="text-align: center;"> <iframe src="/assets/plotly/region_component_space.html" frameborder="0" scrolling="no" height="500px" width="100%" style="border: none;"> </iframe> <figcaption><strong>Figure 5:</strong> Region manifold visualized in region component space.</figcaption> </figure> <p><strong>The cells appear to form a one-dimensional manifold.</strong> In order to interpret this manifold, one can first fit a line to it and visualize it in tissue space. To fit a line, we first connect neighboring cells in this region space using a weighted Delaunay triangulation. Cells closer to one another have higher weights while those further from one another have lower weights. This provides a domain in region space. We can then calculate and eigendecompose the corresponding Laplacian matrix to find the lowest eigenvector which represents the lowest frequency and thus the longest path through the domain. Note that this approach is essentially equivalent to <a href="https://en.wikipedia.org/wiki/Diffusion_map#" rel="external nofollow noopener" target="_blank">diffusion maps</a>, which is a common single-cell trajectory inference method. Again, try interacting with the plot below to confirm that we’ve fit a line along the manifold from above.</p> <figure style="text-align: center;"> <iframe src="/assets/plotly/region_gradient.html" frameborder="0" scrolling="no" height="500px" width="100%" style="border: none;"> </iframe> <figcaption><strong>Figure 6:</strong> Region manifold visualized in region component space and colored by the molecular gradient it describes.</figcaption> </figure> <p><strong>The resulting line forms a gradient</strong> from one tip of the manifold to the other. Now let’s visualize this gradient in the tissue. We can take this same coloring and simply apply it to cells arranged in the tissue domain. Indeed, we find a molecularly-defined gradient that stretches from the center of the tissue outward, matching the way our tissue was constructed.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/region_gradient_tissue.png" alt="" style="width:50%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 7:</strong> Region gradient visualized in tissue space.</figcaption> </figure> <p>However, while this is a really neat way to view region features in the data, one might ultimately wish to <strong>tie this information back to the ground truth region labels</strong> we started with.</p> <h3 id="region-clusters">Region clusters</h3> <p>To assign each cell to a discrete region label, one can cluster cells in region component space. We can simply perform k-means on the top three components we’ve been visualizing so far. We’ll choose $k=4$ given the number of ground truth regions in our simulated dataset. After clustering, we can return to our cells in region component space and color them by our new discrete region labels.</p> <figure style="text-align: center;"> <iframe src="/assets/plotly/region_clusters.html" frameborder="0" scrolling="no" height="500px" width="100%" style="border: none;"> </iframe> <figcaption><strong>Figure 8:</strong> Region manifold visualized in region component space and colored by discrete region clusters.</figcaption> </figure> <p>As expected, it appears that <strong>the resulting clusters divide cells along the gradient</strong>. Once again, we can take this coloring and apply it to our cells in tissue space.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/region_clusters_tissue.png" alt="" style="width:50%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 9:</strong> Region clusters visualized in tissue space.</figcaption> </figure> <p>Ultimately, it seems pretty clear by eye that we’ve identified our ground truth region clusters.</p> <hr> <h2 id="mouse-brain-data">Mouse brain data</h2> <p>Now let’s see how well the above analysis generalizes to real tissues, again using <a href="https://www.nature.com/articles/s41586-021-03705-x" rel="external nofollow noopener" target="_blank">the mouse MOp</a> as an example as in the <a href="/blog/2025/graph-fourier/">last post</a>.</p> <h3 id="region-components-1">Region components</h3> <p>Two gene markers Standing waves along a depth gradient from the top of the tissue to the bottom.</p> <figure style="text-align: center;"> <img src="/assets/figures/regions/mop/region_features.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 10:</strong> Visualization of region components and their gene markers in the tissue.</figcaption> </figure> <h3 id="region-gradients-1">Region gradients</h3> <figure style="text-align: center;"> <img src="/assets/figures/regions/mop/region_gradients.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 11:</strong> Region gradient visualized in region component and tissue spaces.</figcaption> </figure> <h3 id="region-clusters-1">Region clusters</h3> <figure style="text-align: center;"> <img src="/assets/figures/regions/mop/region_clusters.png" alt="" style="width:100%; display: block; margin: 0 auto;"> <figcaption><strong>Figure 11:</strong> Region clusters visualized in terms of their gene markers and in tissue space.</figcaption> </figure> <h2 id="conclusion">Conclusion</h2> </d-article> <d-appendix> <d-footnote-list></d-footnote-list> <d-citation-list></d-citation-list> </d-appendix> <d-bibliography src="/assets/bibliography/2025-05-18-regions.bib"></d-bibliography> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © Copyright 2025 Kamal Maher. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/img-comparison-slider@8.0.6/dist/index.min.js" integrity="sha256-EXHg3x1K4oIWdyohPeKX2ZS++Wxt/FRPH7Nl01nat1o=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/swiper@11.0.5/swiper-element-bundle.min.js" integrity="sha256-BPrwikijIybg9OQC5SYFFqhBjERYOn97tCureFgYH1E=" crossorigin="anonymous"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> </body> </html>