---
layout: distill
title: Defining molecular tissue regions
description: A principled approach to deriving representations of regions in spatial omics data
tags:
giscus_comments: false
date: 2025-05-15
featured: true

authors:
  - name: Kamal Maher
    affiliations:
      name: MIT, Broad Institute

bibliography: 2025-05-18-regions.bib

toc:
  - name: Region components
  - name: Gradients
  - name: Clusters

---

## Region components

As we saw in the [last post](/blog/2025/graph-fourier/), low-pass filtering isolates large-scale patterns.
Recall that low-pass filtering corresponds to multiplying our data matrix by a function of the graph Laplacian matrix.
Let's resume this thought by defining our variables so we are all roughly on the same page.

Our gene expression is given by the cell-by-gene matrix
$$
\mathbf{X} \in \mathbb{R}^{c \times g},
$$
where $c$ is the number of cells and $g$ is the number of genes.

One might think that performing conventional single-cell analysis on low-pass filtered data would identify large-scale transcriptional features over the tissue (as opposed to cell type features).
In this section, we will show that this is indeed the case.

First, we can visualize the result of performing low-pass filtering on a given gene expression signal over the tissue.
Consider a marker gene for one of the inner regions of our tissue.
Move the slider side to side to compare the signal before and after filtering.

<img-comparison-slider>
  {% include figure.liquid path="assets/figures/regions/unfiltered_signal.png" class="img-fluid rounded z-depth-1" slot="first" %}
  {% include figure.liquid path="assets/figures/regions/lowpass_signal.png" class="img-fluid rounded z-depth-1" slot="second" %}
</img-comparison-slider>
<figcaption><strong>Figure 1:</strong> [Interactive] Comparison of a gene expression signal before (left) and after (right) filtering. </figcaption>

Let $\mathbf{\bar X} = [\mathbf{\bar x_1} | ... | \mathbf{\bar x_g}] \in \mathbb{R}^{c \times g}$ represent the cell-by-gene matrix of low-pass filtered gene signals.
The relationships between all pairs of genes can be represented in terms of covariance.
Assuming mean centering, this "low-pass covariance matrix" is given by
$$
\mathbf{\bar C} = \mathbf{\bar X}^{\top} \mathbf{\bar X} \in \mathbb{R}^{g \times g}
$$,
with entry $\mathbf{\bar C}_{ij} = \mathbf{\bar x_i}^{\top} \mathbf{\bar x_j}$ being the similarity between genes $i$ and $j$.
Note that, for quantitative simplicity, we do not perform mean centering and find that it rarely impacts the results.

We can visualize $\mathbf{\bar C}$ as an image.
Gene markers for each region form red blocks along the diagonal, denoting positively covarying groups of genes, i.e. gene programs.
Gene programs describing different regions can be seen to negatively correlate, forming blue blocks along the off-diagonal.
<!-- 
<figure>
  <img src="/assets/img/regions/region_covariance.png" alt="Visualization of low-pass gene-gene covariance and components." style="width:100%">
  <figcaption><strong>Figure 1:</strong> Left: covariance matrix calculated using low-pass filtered gene expression patterns. Right: the corresponding eigenvectors (columns), each of which represents a list of gene weights for a given region component. </figcaption>
</figure> -->


Ultimately, we seek to distill these patterns of complementary gene programs into a simpler representation -- perhaps by assigning each gene a single value.
Let $\mathbf{\bar u} \in \mathbb{R}^{c}$ represent this desired representation.
If $\mathbf{\bar u}$ is a good representation, it should produce a similar covariance pattern to $\mathbf{\bar C}$.
This can be quantified as
$$
\sum_{ij} \mathbf{\bar C}_{ij} (\mathbf{\bar u}_i \mathbf{\bar u}_j).
$$
Thus, we wish to find the embedding $\mathbf{\bar u}$ that maximizes this expression.
<!-- We can do so using the same reasoning as in \textbf{Section \ref{sec:frequencies}}.
First, note that eq. (\ref{eq:quadform_cov_lows}) is a quadratic form and can instead be written as
\begin{equation}
    \mathbf{\bar u}^{\top} \mathbf{\bar C} \mathbf{\bar u}.
\end{equation}
Then the optimal embedding can be written as
\begin{equation} \label{eq:rayleigh_cov_lows}
    \argmax_{\mathbf{\bar u}}{\frac{\mathbf{\bar u}^{\top} \mathbf{\bar C} \mathbf{\bar u}}{\mathbf{\bar u}^{\top} \mathbf{\bar u}}}.
\end{equation}
Finally, just as in eq. (\ref{eq:rayleigh_freq}), the optimal $\mathbf{\bar u}$ is the top eigenvector of $\mathbf{\bar C}$.
We can see this by rearranging the right-hand side of eq. (\ref{eq:rayleigh_cov_lows}) as an eigenvalue problem:
\begin{align} \label{eq:eig_cov_lows}
    & w = {\frac{\mathbf{\bar u}^{\top} \mathbf{\bar C} \mathbf{\bar u}}{\mathbf{\bar u}^{\top} \mathbf{\bar u}}} \nonumber \\
    &\rightarrow w \mathbf{\bar u}^{\top} \mathbf{\bar u} = \mathbf{\bar u}^{\top} \mathbf{\bar C} \mathbf{\bar u} \\
    &\rightarrow w \mathbf{\bar u} = \mathbf{\bar C} \mathbf{\bar u}. \nonumber
\end{align}
However, all of the top eigenvectors should represent gene programs underlying a prominent and unique pattern.
This is because they capture the most information while also being orthogonal to the other eigenvectors.
Note that this process is simply PCA.
% However, we derive it from first principles for clarity and also to make clearer its interpretation as "frequencies in data space". -->

<!-- We denote the full eigenbasis as $\mathbf{\bar U} = [\mathbf{\bar u}_1 | \mathbf{\bar u}_g] \in \mathbb{R}^{g \times g}$.
We can also visualize $\mathbf{\bar U}$ as an image (\textbf{Figure \ref{fig:cov_lowpass}B}).
Each column of the image represents a component $\mathbf{\bar u}_i$.
While the top eigenvector contains entirely negative values and likely corresponds to a scaling factor, the following three components all contain complimentary gene programs from different regions.
Component 2 described the relationship between the outermost and second innermost regions (positive and negative, respectively).
Component 3 described the relationship between the second outermost and innermost regions (positive and negative, respectively).
Component 4 described the relationship between the second innermost and innermost regions (positive and negative, respectively).
Further components did not appear to involve region marker genes. -->

<!-- While the above analysis identifies related gene programs, one can visualize these programs in the tissue by projecting cells into the corresponding spaces.
For instance, given component $i$, one can project each cell into this gene program space by calculating
\begin{equation}
    \mathbf{\bar s}_i = \mathbf{\bar X} \mathbf{\bar u}_i \in \mathbb{R}^c.
\end{equation}
One can then visualize this component in the tissue (\textbf{Figure \ref{fig:region_features_sim}A}).
The gene weights defining each component are given by $\mathbf{s}_i$.
Thus, for a given component, the gene markers for the corresponding positive (negative) gene program are the same as those that are red (blue) in \textbf{Figure \ref{fig:cov_lowpass}B}.
While the information is the same, the visualization in \textbf{Figure \ref{fig:region_features_sim}B} is more compact and will be used throughout the remainder of this work.
Note that the components with more cells involved (e.g. component 2) are ranked higher than those with fewer cells involved (e.g. component 4).
While this may be obvious given the definition of covariance in general, it has a special significance for spatial analyses, as larger-scale patterns will likely be prioritized over smaller-scale ones.

\begin{figure}[t]
\centering\includegraphics[width=\textwidth]{MIT-thesis-template/chapter3_figs/simulations/region_features.png} 
\caption{
\textbf{Region components.}
\textbf{A)} Components visualized in tissue space.
\textbf{B)} Gene markers (low-pass filtered) for each component, i.e. most positive and most negative loadings.
Markers for each side of the pattern (positive and negative) are plotted on the right in the same order as listed on the left.
Alternative visualization of the information provided in Figure \ref{fig:cov_lowpass}B.
\label{fig:region_features_sim}
}
\end{figure} -->


## Gradients

The above components indeed appear to contain orthogonal information.
They each describe unique sets of regions.
Together, however, these features can describe more complex region patterns.
Rather than projecting cells onto a single component, one can instead project onto multiple components that form a multidimensional "region space".
For instance, we can project cells onto the space defined by the top three components by calculating
$$
\mathbf{\bar S}_{:3} = \mathbf{\bar X} \mathbf{\bar U}_{:3} \in \mathbb{R}^{c \times 3}
$$
where $\mathbf{\bar U}_{:3} = [\mathbf{\bar u}_1 | \mathbf{\bar u}_2 | \mathbf{\bar u}_3]$.
Cells can then be plotted in the resulting three-dimensional space Figure.

Note that the cells appear to form a one-dimensional manifold.
In order to interpret this manifold, one can first fit a line to it and visualize it in tissue space.
To fit a line, we first connect neighboring cells in this region space using a weighted Delaunay triangulation.
Cells closer to one another have higher weights while those further from one another have lower weights.
This provides a \textit{domain} in region space.
We can then calculate and eigendecompose the corresponding Laplacian matrix to find the lowest eigenvector which represents the lowest frequency and thus the longest path through the domain.
The result is the coloring shown in Figure.
Note that this approach is essentially the same as diffusion maps, which is a conventional single-cell trajectory inference method.

<iframe src="{{ '/assets/plotly/region_gradient.html' | relative_url }}"
        frameborder='0'
        scrolling='no'
        height="500px"
        width="100%"
        style="border: none; display: block; margin: 0 auto;"></iframe>
<figcaption><strong>Figure 3:</strong> [Interactive] Molecular gradient visualized in region space. </figcaption>

The resulting coloring reveals a continuous, molecularly defined "depth" axis, as seen when plotted in the tissue.
We could alternatively make sense of this manifold given the components plotted in.
Component 1 describes an "inside-outside" pattern with only the outermost layer considered outside.
Component 2 then splits the inside layers from component 1 in the same way.
Finally, component 3 further splits the inside layers from component 2.
Thus, together, these features can distinguish each layer, while the inside-outside patterns enable the layers to be related to one another in terms of depth.

<figure>
  <img src="/assets/figures/regions/region_gradient.png" alt="" style="width:100%">
  <figcaption><strong>Figure 2:</strong> Molecular gradient visualized in the tissue. </figcaption>
</figure>

## Clusters