---
layout: distill
title: "Spatial Omics II: Defining Multicellular Regions"
description: A principled approach to representing multicellular regions in spatial omics data
tags:
giscus_comments: false
date: 2025-05-18
featured: false
categories: spatial-omics
thumbnail: assets/figures/regions/region_gradient.png

authors:
  - name: Kamal Maher
    affiliations:
      name: MIT, Broad Institute

bibliography: 2025-05-18-regions.bib

toc:
  - name: Introduction
  - name: Simulation
    subsections:
      - name: Simulated region components
      - name: Simulated region gradients
      - name: Simulated region clusters
  - name: Mouse brain
    subsections:
      - name: Brain region components
      - name: Brain region gradients
      - name: Brain region clusters
  - name: Human lymph node
    subsections:
      - name: Lymph node region components
      - name: Lymph node region clusters
  - name: Conclusion

images:
  compare: true
  slider: true
---

<style>
  .slider-with-shadows {
    --default-handle-shadow: 0px 0px 5px rgba(0, 0, 0, 1);
    --divider-shadow: 0px 0px 5px rgba(0, 0, 0, 0.5);
  }
</style>


## Introduction

As we saw in the [last post](/blog/2025/graph-fourier/), low-pass filtering isolates large-scale patterns.
Recall that low-pass filtering corresponds to multiplying our data matrix by a function of the graph Laplacian matrix.
Let's resume this thought by defining our variables so we are all roughly on the same page.

Our gene expression is given by the cell-by-gene matrix

$$
\mathbf{X} \in \mathbb{R}^{c \times g},
$$

where $c$ is the number of cells and $g$ is the number of genes.

One might think that performing conventional single-cell analysis on low-pass filtered data would identify large-scale transcriptional features over the tissue (as opposed to cell type features).
In this section, we will show that this is indeed the case.

First, we can visualize the result of performing low-pass filtering on a given gene expression signal over the tissue.
Consider a marker gene for one of the inner regions of our tissue.
Drag the slider left to right to compare the signal before and after filtering.

<div style="width: 50%; max-width: 768px; margin: 0 auto;">
  <img-comparison-slider class="slider-with-shadows">
    {% include figure.liquid path="assets/figures/fourier/tissue_before_filtering.png" class="img-fluid rounded z-depth-1" slot="first" %}
    {% include figure.liquid path="assets/figures/fourier/tissue_after_lowpass.png" class="img-fluid rounded z-depth-1" slot="second" %}
  </img-comparison-slider>
</div>
<figcaption><strong>Figure 1:</strong> Comparison of a gene expression signal before (left image) and after (right image) low-pass filtering. </figcaption>
<br>

Note that the small-scale variation in expression has been washed away.
Neighboring cells are forced to look more similar to one another, in effect isolating only *large*-scale patterns.

---


## Simulation

### Simulated region components

Now that we've isolated large-scale gene expression patterns, we can start to **compare them to find interesting gene-gene relationships**.
Intuitively, a group of gene expression patterns that overlap in the tissue should represent a region.
We can find these groups by looking at pairwise relationships between gene signals.
Consider low-pass filtered gene signals $\mathbf{\bar x}_i, \mathbf{\bar x}_j \in \mathbb{R}^n$.
Because these signals are vectors and we want a scalar measure of similarity, we could just compare them by taking the inner product:

$$
\mathbf{\bar x}_i^{\top} \mathbf{\bar x}_j \in \mathbb{R}.
$$

With some mean centering, this is defined as the covariance.
While we will continue to refer to this as covariance, we will omit all mean centering for simplicity (although it will often add a "junk component" as we will see in a moment).

But we aren't just interested in one pair of genes; **we want to look at all pairwise relationships**.
Let $\mathbf{\bar X} = [\mathbf{\bar x}_1 | ... | \mathbf{\bar x}_g] \in \mathbb{R}^{n \times g}$ represent the cell-by-gene matrix of low-pass filtered gene signals.
Then the gene-by-gene covariance matrix is given by

$$
\mathbf{C} = \mathbf{\bar X}^{\top} \mathbf{\bar X} \in \mathbb{R}^{g \times g},
$$

with entries $\mathbf{C}_{ij} = \mathbf{\bar x}_i^{\top} \mathbf{\bar x}_j$.
We can visualize $\mathbf{C}$ as a heatmap.
The genes are sorted based on their corresponding ground truth region patterns, so we expect to see interesting groups of covarying genes as blocks.
(If you look closely, you'll see that these blocks are 4x4, as there are four gene markers for each pattern.)
For instance, gene markers for each region form red blocks along the diagonal, denoting positively covarying groups of genes, i.e. gene programs.
Furthermore, different blocks appear to negatively covary, forming blue blocks along the *off*-diagonal.
Conceptually, this just means that gene markers for each region are mutually exclusive.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/region_covariance.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 2:</strong> Visualization of the low-pass covariance matrix.</figcaption>
</figure>

Ultimately, **we seek to distill these blocks into a simpler representation** -- perhaps by grouping genes within related blocks into a single "gene program" describing a region.
It turns out we can do this by eigendecomposing $\mathbf{C}$, i.e. performing PCA.
(For a primer on PCA, see [this post](/blog/2025/pca).)
Briefly, eigendecomposing the gene-gene covariance matrix yields the eigenbasis

$$
\mathbf{U} = [ \mathbf{u}_1 | ... | \mathbf{u}_g ] \in \mathbb{R}^{g \times g},
$$

where $\mathbf{u}_i \in \mathbb{R}^g$ represents the gene loadings for PC$i$.
In other words, it describes each gene's participation in gene program $i$.
We can visualize this matrix as a heatmap as well.
The rows represent each gene in the same order as in Figure 2.
The columns now represent PCs, or "region components", and the colors describe how much and in what direction a given gene contributes to a given component.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/region_components.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 3:</strong> Visualization of low-pass gene programs, i.e. "region components". The right-hand plot is an enlarged version of the inset in the top left corner of the left-hand plot.</figcaption>
</figure>

The first component is entirely negative and is likely just a consequence of neglecting mean centering (the "junk component" mentioned above).
The second, third, and fourth components each appear to describe relationships between regions, with each representing a positive and negative region.

We can also visualize each component in the tissue by projecting cells onto each gene program, i.e. $\mathbf{X} \mathbf{U} \in \mathbb{R}^{n \times g}$.
We end up seeing that they each describe two regions of the tissue.
The first component describes the outer ring versus everything inside of it, the second component describes the second-most outer right versus everything inside, and the third component describes the third-most outer ring versus everything inside.
We can also sort gene loadings to identify gene markers for each region component.
This gives a set of top genes describing the "positive region" and a different set for the "negative region".
We can visualize each of these markers in the tissue as well, seeing that they are indeed expressed within the regions they describe.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/region_components_tissue.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 4:</strong> Visualization of region components and their gene markers in the tissue.</figcaption>
</figure>

Altogether, we find that **each component represents a large-scale pattern between regions**.
Note that the gene marker information shown in Figure 4 is the same as the information shown in Figure 3, just in a different form.


### Simulated region gradients

The above components indeed appear to contain orthogonal information, as they each describe unique sets of regions.
**Together, however, these features might describe more complex region patterns**.
Rather than projecting cells onto a single component, one can instead project onto multiple components that form a multidimensional "region space".
In other words, rather than plotting cells in the tissue and *coloring* by the top three different region components as above, we can directly plot each cell in three-dimensional component space and see what shapes they form.
Try interacting with the plot below to get a sense of the resulting shape.

<figure style="text-align: center;">
  <iframe src="{{ '/assets/plotly/region_component_space.html' | relative_url }}"
          frameborder='0'
          scrolling='no'
          height="500px"
          width="100%"
          style="border: none;">
  </iframe>
  <figcaption><strong>Figure 5:</strong> Region manifold visualized in region component space.</figcaption>
</figure>

**The cells appear to form a one-dimensional manifold.**
In order to interpret this manifold, one can first fit a line to it and visualize it in tissue space.
To fit a line, we first connect neighboring cells in this region space using a weighted Delaunay triangulation.
Cells closer to one another have higher weights while those further from one another have lower weights.
This provides a domain in region space.
We can then calculate and eigendecompose the corresponding Laplacian matrix to find the lowest eigenvector which represents the lowest frequency and thus the longest path through the domain.
Note that this approach is essentially equivalent to [diffusion maps](https://en.wikipedia.org/wiki/Diffusion_map#), which is a common single-cell trajectory inference method.
Again, try interacting with the plot below to confirm that we've fit a line along the manifold from above.

<figure style="text-align: center;">
  <iframe src="{{ '/assets/plotly/region_gradient.html' | relative_url }}"
          frameborder='0'
          scrolling='no'
          height="500px"
          width="100%"
          style="border: none;">
  </iframe>
  <figcaption><strong>Figure 6:</strong> Region manifold visualized in region component space and colored by the molecular gradient it describes.</figcaption>
</figure>

**The resulting line forms a gradient** from one tip of the manifold to the other.
Now let's visualize this gradient in the tissue.
We can take this same coloring and simply apply it to cells arranged in the tissue domain.
Indeed, we find a molecularly-defined gradient that stretches from the center of the tissue outward, matching the way our tissue was constructed.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/region_gradient_tissue.png"
       alt=""
       style="width:50%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 7:</strong> Region gradient visualized in tissue space.</figcaption>
</figure>

However, while this is a really neat way to view region features in the data, one might ultimately wish to **tie this information back to the ground truth region labels** we started with.


### Simulated region clusters

To assign each cell to a discrete region label, one can cluster cells in region component space.
We can simply perform k-means on the top three components we've been visualizing so far.
We'll choose $k=4$ given the number of ground truth regions in our simulated dataset.
After clustering, we can return to our cells in region component space and color them by our new discrete region labels.

<figure style="text-align: center;">
  <iframe src="{{ '/assets/plotly/region_clusters.html' | relative_url }}"
          frameborder='0'
          scrolling='no'
          height="500px"
          width="100%"
          style="border: none;">
  </iframe>
  <figcaption><strong>Figure 8:</strong> Region manifold visualized in region component space and colored by discrete region clusters.</figcaption>
</figure>

As expected, it appears that **the resulting clusters partition cells along the gradient**.
Once again, we can take this coloring and apply it to our cells in tissue space.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/region_clusters_tissue.png"
       alt=""
       style="width:50%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 9:</strong> Region clusters visualized in tissue space.</figcaption>
</figure>

Ultimately, it seems by eye that we've identified our ground truth region clusters.

---


## Mouse brain

Now let's see how well the above analysis generalizes to real tissues, using [the mouse MOp](https://www.nature.com/articles/s41586-021-03705-x) as an example like in the [last post](/blog/2025/graph-fourier/).
Since it's the inspiration for our simulation, we can just repeat each of the above analyses to identify region components, gradients, and clusters.


### Brain region components

Recall our previous filtering results

Two gene markers
Standing waves along a depth gradient from the top of the tissue to the bottom.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/mop/region_features.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 10:</strong> Visualization of region components and their gene markers in the tissue.</figcaption>
</figure>


### Brain region gradients

Unfortunately, the amount of data prohibits interactive 3D visualization.
But we can still plot the cells from multiple angles to get a sense of their latent shape.
Like in our simulation, it seems they form a one-dimensional manifold.
We can apply the same approach as above to identify a gradient, which visibly follows the manifold from end to end.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/mop/region_gradients.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 11:</strong> Region gradient visualized in region component and tissue spaces.</figcaption>
</figure>

When plotted in the tissue, the gradient appears to run between the outer and inner layers of the neocortex.
Thus, it appears our gradient identification approach applies to real biological data as well.
However, to make the anatomical layout of this data clearer, we can assign discrete region labels and see if they reflect the known anatomy.


### Brain region clusters

Just as in our simulation, we can apply k-means to the top three components visualized above to assign each cell a region label.
Unlike in our simulation, however, we don't know the ground truth number of regions that should inform our choice of $k$.
Nevertheless, we can approximate it based on the [known anatomy](https://atlas.brain-map.org/atlas?atlas=1&plate=100960348), in which four cortical layers named L2/3, L4, L5, and L6 are sandwiched between the meninges above and white matter (WM) below.
Thus, we will set $k=6$, perform k-means, and expect to find a similar set of regions.

One way to compare our region clusters to the known anatomy is based on their gene markers.
Using [Scanpy's standard one-vs-all t-test-based approach](), we can identify the top three gene markers for each region and compare them to the known anatomy with references such as [CELLxGENE](https://cellxgene.cziscience.com/).
This results in a simple annotation of each region that matches the literature quite well.
Additionally, note that the markers for one region appear to bleed into the next region.
This is likely due to the low-pass nature of the data, in which the boundaries between regions are blurred.
Finally, we can plot the resulting region labels in the tissue domain to find that they form layers characteristic of the neocortex, further bolstering our approach.

<figure style="text-align: center;">
  <img src="/assets/figures/regions/mop/region_clusters.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 12:</strong> Region clusters visualized in terms of their gene markers and in tissue space.</figcaption>
</figure>


--- 


## Human lymph node

The above analysis is by no means restricted to tissues with layered structures.
It should, in principle, be applicable to any tissue with any arrangement of molecularly-distinct regions.
To test that, we will use a new dataset that was collected from a different organ, from a different species, and using a different spatial transcriptomic technology.
Our new dataset is a sample of the healthy human lymph node and consists of 377 genes measured across 377,957 cells measured using [Xenium](https://www.biorxiv.org/content/10.1101/2023.12.07.570603v2).
This dataset will also lay the groundwork for [the next post on intercellular interactions]().
<!-- Because it does not have a known layered structure, we will focus briefly on identifying region components and clusters. -->

Intuition for the organ
Intuition for filtering (slider)


### Lymph node region components

<figure style="text-align: center;">
  <img src="/assets/figures/regions/hln/region_features.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 13:</strong> Visualization of region components and their gene markers in the tissue.</figcaption>
</figure>


### Lymph node region clusters

<figure style="text-align: center;">
  <img src="/assets/figures/regions/hln/region_clusters.png"
       alt=""
       style="width:100%; display: block; margin: 0 auto;">
  <figcaption><strong>Figure 14:</strong> Region clusters visualized in terms of their gene markers and in tissue space.</figcaption>
</figure>

---


## Conclusion